postgres/datadog_checks/postgres/autodiscovery-script.py

```diff
@@ -0,0 +1,193 @@
+
+class PostgresAutodiscovery(Discovery): 
+    def __init__(self, host: str, max_conn: int) -> None:
+        super(PostgresAutodiscovery, self).__init__(self._get_databases, include={'.*': 10}, exclude=[], interval=60)
+        self.host = host
+        relations_config = [{'relation_regex': '.*'}]
+        self._relations_manager = RelationsManager(relations_config)
+        self._conn_pool = MultiDatabaseConnectionPoolLimited(self._connect, max_conn)
+        self.default_ttl = 60000
+        # get once to cache dbs
+        self.get_items()
+        # self._conn_pool = psycopg2.ThreadedConnectionPool(minconn=0, maxconn=maxconn, user="postgres", password="p0stgres", port="5432")
+    
+    def get_items(self) -> List[str]:
+        """
+        Get_items() from parent class returns a generator with four objects:
+            > yield pattern, key(item), item, config
+        This function takes the item of interest (dbname) from this four-tuple
+        and returns the full list of database names from the generator.
+        """
+        items = list(super().get_items())
+        items_parsed = [item[1] for item in items]
+        return items_parsed
+   
+    def _connect(self, dbname: str = None) -> None:
+        # Use ident method
+        connection_string = "host="+self.host+" user=postgres password=p0stgres"
+        if dbname is not None:
+            connection_string += " dbname=" + dbname
+        conn = psycopg2.connect(connection_string)
+
+        # print("connected")
+        return conn
+
+    def _get_autodiscovery_query(self) -> str:
+        autodiscovery_query = AUTODISCOVERY_QUERY.format(columns=', '.join(['datname']))
+        return autodiscovery_query
+
+    def _get_databases(self) -> List[str]:
+        conn = self._conn_pool.get_connection('postgres', self.default_ttl)
+        cursor = conn.cursor()
+        autodiscovery_query = self._get_autodiscovery_query()
+        cursor.execute(autodiscovery_query)
+        databases = list(cursor.fetchall())
+        databases = [x[0] for x in databases] # fetchall returns list of tuples representing rows, so need to parse
+        print("Databases found were: ", databases)
+        return databases 
+    
+    def run_query_scope(self, cursor, scope, cols, descriptors):
+        # try:
+        query = fmt.format(scope['query'], metrics_columns=", ".join(cols))
+        schema_field = get_schema_field(descriptors)
+        # print(schema_field)
+        formatted_query = self._relations_manager.filter_relation_query(query, schema_field)
+        # print(formatted_query)
+        # while(1):
+        #     pass
+        cursor.execute(formatted_query)
+        
+        results = cursor.fetchall()
+        return results
+
+    def query_relations(self, database: str) -> None:
+        # print(cached_dbs)
+        conn = None
+        # wait for connection to open up
+        while conn == None:
+            conn = self._conn_pool.get_connection(database, self.default_ttl)
```

Comment:
I don't know the answer, but is it better to make this a blocking call rather than looping indefinitely `while conn == True`